src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UsingStatement.cs

```diff
@@ -288,20 +305,25 @@ private BoundStatement RewriteUsingStatementTryFinally(SyntaxNode syntax, BoundB
                 disposedExpression = local;
             }
 
-            // local.Dispose()
             BoundExpression disposeCall;
-
-            MethodSymbol disposeMethodSymbol;
-            if (Binder.TryGetSpecialTypeMember(_compilation, SpecialMember.System_IDisposable__Dispose, syntax, _diagnostics, out disposeMethodSymbol))
+            if (awaitOpt == null && Binder.TryGetSpecialTypeMember(_compilation, SpecialMember.System_IDisposable__Dispose, syntax, _diagnostics, out MethodSymbol disposeMethodSymbol))
             {
+                // local.Dispose()
                 disposeCall = BoundCall.Synthesized(syntax, disposedExpression, disposeMethodSymbol);
             }
+            else if (awaitOpt != null && TryGetWellKnownTypeMember(syntax, WellKnownMember.System_IAsyncDisposable__DisposeAsync, out MethodSymbol disposeAsyncMethodSymbol))
```

Comment:
If we're reporting use-site diagnostics here, could we skip reporting the use-site diagnostic for `IAsyncDisposable` in binding?

Reply:
The checks here don't handle missing `IDisposable`, only missing `IDisposable.Dipose()`.
I agree with you that it's weird to split the use-site diagnostics between here and initial binding. I'll look into this more.

---
In reply to: [160593496](https://github.com/dotnet/roslyn/pull/23961#discussion_r160593496) [](ancestors = 160593496)

Reply:
Doesn't `TryGetWellKnownTypeMember` report a diagnostic if the type is missing?

Reply:
If I remove the reporting of use-site diagnostics in binding (for missing `IDisposable`), then execution does not reach lowering, even when I ask for emit diagnostics.
I think that's because there are errors reported from binding anyways. Namely:
```
                // (6,16): error CS1674: 'C': type used in a using statement must be implicitly convertible to 'System.IDisposable'
                //         using (new C())
                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "new C()").WithArguments("C").WithLocation(6, 16),
                // (9,16): error CS1674: 'C': type used in a using statement must be implicitly convertible to 'System.IDisposable'
                //         using (var x = new C())
                Diagnostic(ErrorCode.ERR_NoConvToIDisp, "var x = new C()").WithArguments("C").WithLocation(9, 16)
```

---
In reply to: [160863521](https://github.com/dotnet/roslyn/pull/23961#discussion_r160863521) [](ancestors = 160863521)

Reply:
I believe all error reporting for missing members and such should be done in binding. IIRC, the local rewriter can assume that these exist here. Example in: [https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StackAlloc.cs](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StackAlloc.cs) when getting the span type

---
In reply to: [161024454](https://github.com/dotnet/roslyn/pull/23961#discussion_r161024454) [](ancestors = 161024454,160863521)

Reply:
Initial binding doesn't currently check for existing of `IDisposable.Dispose()`.
There are other examples of well-known types and members which are only checked during rewriting, such as `AsyncRewriter.VerifyPresenceOfRequiredAPIs()`. 

I don't see anything wrong with that, especially since getting in a situation where `IDisposable.Dispose()` is missing, but `IDisposable` exists wouldn't seem common. The same applies to async counterpart APIs.